File: 2026-05/2026-05-14/mirrorImage.js
function isMirrorImage(str1, str2) {
  const SYMMETRIC_CHARACTERS = ["W", "T", "Y", "U", "I", "O", "H", "A", "X", "V", "M", "w", "o", "x", "v", "0", "8", "=", "+", ":", "|", "-", "_", "*", "^", "!", ".", " "];
  const MIRRORED_PAIRS = {
    "[": "]",
    "{": "}",
    "<": ">",
    "b": "d",
    "p": "q",
    "(": ")"
  };

  const isMirrorable = str => {
    for (const character of str) {
      if (!SYMMETRIC_CHARACTERS.includes(character) && !Object.keys(MIRRORED_PAIRS).includes(character) && !Object.values(MIRRORED_PAIRS).includes(character)) {
        return false;
      }
    }

    return true;
  }

  if (!isMirrorable(str1) || !isMirrorable(str2)) {
    return false;
  }

  return str1
    .split("")
    .map(character => {
      let mirroredCharacter = character;

      for (const [char, swap] of Object.entries(MIRRORED_PAIRS)) {

        if (char === character) {
          mirroredCharacter = swap;
          break;
        }

        if (swap === character) {
          mirroredCharacter = char;
          break;
        }
      }

      return mirroredCharacter;
    })
    .reverse()
    .join("")
    === str2;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. isMirrorImage("[HOW]", "[WOH]") should return true.
2. isMirrorImage("MOM", "MOM") should return true.
3. isMirrorImage("vow", "wov") should return true.
4. isMirrorImage("TIM", "TIM") should return false.
5. isMirrorImage("{WOW}", "}WOW{") should return false.
6. isMirrorImage("XXVII", "IIV%X") should return false.
7. isMirrorImage("><(((*>", "<*)))><") should return true.
8. isMirrorImage("WTYUIOHAXVMwoxv08=+:|-_*^!.[]{}<>bdpq()", "()pqbd<>{}[].!^*_-|:+=80vxowMVXAHOIUYTW") should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 05 May/05-14/mirror_image.py
def is_mirror_image(s1, s2):
    SYMMETRIC_CHARACTERS = ["W", "T", "Y", "U", "I", "O", "H", "A", "X", "V", "M", "w", "o", "x", "v", "0", "8", "=", "+", ":", "|", "-", "_", "*", "^", "!", ".", " "]
    MIRRORED_PAIRS = {
        "[": "]",
        "{": "}",
        "<": ">",
        "b": "d",
        "p": "q",
        "(": ")"
    }

    def is_mirrorable(s):
        for character in s:
            if character not in SYMMETRIC_CHARACTERS and character not in MIRRORED_PAIRS.keys() and character not in MIRRORED_PAIRS.values():
                return False

        return True

    if not is_mirrorable(s1) or not is_mirrorable(s2):
        return False

    def mirror_swap(character):
        mirrored_character = character

        for char, swap in MIRRORED_PAIRS.items():
            if char == character:
                mirrored_character = swap
                break

            if swap == character:
                mirrored_character = char
                break

        return mirrored_character

    return "".join(reversed([mirror_swap(character) for character in s1])) == s2

# --- TEST SUITE ---

tests_text = r'''
1. is_mirror_image("[HOW]", "[WOH]") should return True.
2. is_mirror_image("MOM", "MOM") should return True.
3. is_mirror_image("vow", "wov") should return True.
4. is_mirror_image("TIM", "TIM") should return False.
5. is_mirror_image("{WOW}", "}WOW{") should return False.
6. is_mirror_image("XXVII", "IIV%X") should return False.
7. is_mirror_image("><(((*>", "<*)))><") should return True.
8. is_mirror_image("WTYUIOHAXVMwoxv08=+:|-_*^!.[]{}<>bdpq()", "()pqbd<>{}[].!^*_-|:+=80vxowMVXAHOIUYTW") should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)